pydis_site/apps/api/serializers.py

```diff
@@ -249,15 +256,69 @@ class Meta:
         fields = ('id', 'name', 'colour', 'permissions', 'position')
 
 
-class UserSerializer(BulkSerializerMixin, ModelSerializer):
+class UserListSerializer(ListSerializer):
+    """List serializer for User model to handle bulk updates."""
+
+    def create(self, validated_data: list) -> list:
+        """Override create method to optimize django queries."""
+        new_users = []
+        seen = set()
+
+        for user_dict in validated_data:
+            if user_dict["id"] in seen:
+                raise ValidationError(
+                    {"id": f"User with ID {user_dict['id']} given multiple times."}
+                )
+            seen.add(user_dict["id"])
+            new_users.append(User(**user_dict))
+
+        return User.objects.bulk_create(new_users, ignore_conflicts=True)
+
+    def update(self, instance: QuerySet, validated_data: list) -> list:
+        """
+        Override update method to support bulk updates.
+
+        ref:https://www.django-rest-framework.org/api-guide/serializers/#customizing-multiple-update
+        """
+        instance_mapping = {user.id: user for user in instance}
+
+        updated = []
+        fields_to_update = set()
+        for user_data in validated_data:
+            for key in user_data:
+                fields_to_update.add(key)
+
+                try:
+                    user = instance_mapping[user_data["id"]]
+                except KeyError:
+                    raise NotFound({"id": f"User with id {user_data['id']} not found."})
```

Comment:
Only `ValidationErrors` use the convention of the field name as the key. 404s raised by DRF elsewhere use `detail` as the key.
```suggestion
                    raise NotFound({"detail": f"User with id {user_data['id']} not found."})
```

Reply:
I'm adding this comment for posterity: the reason it's okay to raise a 404 here and risk failing the entire request is because the syncer shouldn't ever run into a race condition where it's trying to update a user that has been deleted (or not created yet). The former is impossible because it never deletes users; it only sets `in_guild` as false. The latter is impossible because if it sees the user doesn't exist in the DB, it will try to create rather than update.